apps/mymedicare_cb/models.py

```diff
@@ -1,76 +1,106 @@
-                    first_name=user_info['given_name'],
-                    last_name=user_info['family_name'],
-                    email=user_info['email'])
-        user.set_unusable_password()
-        user.save()
-    UserProfile.objects.get_or_create(
-        user=user, user_type='BEN')
-    group = Group.objects.get(name='BlueButton')
-    user.groups.add(group)
-    # Log in the user
-    user.backend = 'django.contrib.auth.backends.ModelBackend'
-
-    # Determine patient_id
-    fhir_source = get_resourcerouter()
-    crosswalk, _ = Crosswalk.objects.get_or_create(
-        user=user, fhir_source=fhir_source)
-    hicn = user_info.get('hicn', "")
-    crosswalk.set_hicn(hicn)
-    crosswalk.save()
+        pass
+
+    first_name = user_info.get('given_name', "")
+    last_name = user_info.get('family_name', "")
+    email = user_info.get('email', "")
 
     try:
-        backend_data = authenticate_crosswalk(crosswalk)
+        fhir_id, backend_data = match_hicn_hash(hicn_hash)
         # Get first and last name from FHIR if not in OIDC Userinfo response.
-        if user_info['given_name'] == "" or user_info['family_name'] == "":
-            if 'entry' in backend_data:
-                if 'name' in backend_data['entry'][0]['resource']:
-                    names = backend_data['entry'][0]['resource']['name']
-                    first_name = ""
-                    last_name = ""
-                    for n in names:
-                        if n['use'] == 'usual':
-                            last_name = n['family']
-                            first_name = n['given'][0]
-                        if last_name or first_name:
-                            user.first_name = first_name
-                            user.last_name = last_name
-                            user.save()
-    except (UpstreamServerException, exceptions.NotFound):
-        logger.error("Failed to connect Beneficiary "
-                     "to FHIR")
+        if first_name == "" or last_name == "":
+            first_name, last_name = extract_beneficiary_names(backend_data)
+    except exceptions.NotFound:
+        fhir_id = None
 
+    fhir_source = get_resourcerouter()
+
+    user = create_beneficiary_record(username=subject,
+                                     user_id_hash=hicn_hash,
+                                     fhir_id=fhir_id,
+                                     fhir_source=fhir_source,
+                                     first_name=first_name,
+                                     last_name=last_name,
+                                     email=email)
     return user
 
 
+def create_beneficiary_record(username=None,
+                              user_id_hash=None,
+                              fhir_id=None,
+                              fhir_source=None,
+                              first_name=None,
+                              last_name=None,
+                              email=None):
+    with transaction.atomic():
```

Comment:
Can username, user_id_hash, fhir_id be None? Suggest that we throw an exception if this happens.

Reply:
added assertions for `not None`